klippy/extras/exclude_object.py

```diff
@@ -0,0 +1,302 @@
+            offset[3] += self.extruder_adj
+            self.extruder_adj = 0
+
+        if offset[2] != 0 and newpos[2] != self.last_position_excluded[2]:
+            offset[2] = 0
+
+        if self.extruder_adj != 0 and \
+            newpos[3] != self.last_position_excluded[3]:
+            offset[3] += self.extruder_adj
+            self.extruder_adj = 0
+
+        tx_pos = newpos[:]
+        for i in range(4):
+            tx_pos[i] = newpos[i] - offset[i]
+        self.next_transform.move(tx_pos, speed)
+
+    def _ignore_move(self, newpos, speed):
+        offset = self._get_extrusion_offsets()
+        for i in range(3):
+            offset[i] = newpos[i] - self.last_position_extruded[i]
+        offset[3] = offset[3] + newpos[3] - self.last_position[3]
+        self.last_position[:] = newpos
+        self.last_position_excluded[:] =self.last_position
+        self.max_position_excluded = max(self.max_position_excluded, newpos[3])
+
+    def _move_into_excluded_region(self, newpos, speed):
+        self.in_excluded_region = True
+        self._ignore_move(newpos, speed)
+
+    def _move_from_excluded_region(self, newpos, speed):
+        self.in_excluded_region = False
+
+        # This adjustment value is used to compensate for any retraction
+        # differences between the last object printed and excluded one.
+        self.extruder_adj = self.max_position_excluded \
+            - self.last_position_excluded[3] \
+            - (self.max_position_extruded - self.last_position_extruded[3])
+        self._normal_move(newpos, speed)
+
+    def _test_in_excluded_region(self):
+        # Inside cancelled object
+        return self.current_object in self.excluded_objects \
+            and self.initial_extrusion_moves == 0
+
+    def get_status(self, eventtime=None):
+        status = {
+            "objects": self.objects,
+            "excluded_objects": self.excluded_objects,
+            "current_object": self.current_object
+        }
+        return status
+
+    def move(self, newpos, speed):
+        move_in_excluded_region = self._test_in_excluded_region()
+        self.last_speed = speed
+
+        if move_in_excluded_region:
+            if self.in_excluded_region:
+                self._ignore_move(newpos, speed)
+            else:
+                self._move_into_excluded_region(newpos, speed)
+        else:
+            if self.in_excluded_region:
+                self._move_from_excluded_region(newpos, speed)
+            else:
+                self._normal_move(newpos, speed)
+
+    cmd_EXCLUDE_OBJECT_START_help = "Marks the beginning the current object" \
+                                    " as labeled"
+    def cmd_EXCLUDE_OBJECT_START(self, gcmd):
+        name = gcmd.get('NAME').upper()
+        if not any(obj["name"] == name for obj in self.objects):
+            self._add_object_definition({"name": name})
+        self.current_object = name
+        self.was_excluded_at_start = self._test_in_excluded_region()
```

Comment:
Stale attribute?  It doesn't appear that `was_excluded_at_start` is accessed in the module.